src/main/java/org/dependencytrack/persistence/ComponentQueryManager.java

```diff
@@ -156,56 +159,178 @@ public PaginatedResult getComponents(final Project project, final boolean includ
-            querySring +=
-                    " && !(" +
-                            " SELECT FROM org.dependencytrack.model.RepositoryMetaComponent m " +
-                            " WHERE m.name == this.name " +
-                            " && m.namespace == this.group " +
-                            " && m.latestVersion != this.version " +
-                            " && this.purl.matches('pkg:' + m.repositoryType.toString().toLowerCase() + '/%') " +
-                            " ).isEmpty()";
+            queryString +=
+                    """
+                        AND NOT (NOT EXISTS (
+                        SELECT "M"."ID" AS "DN_APPID"
+                        FROM "REPOSITORY_META_COMPONENT" "M" WHERE "M"."NAME" = "A0"."NAME"
+                        AND "M"."NAMESPACE" = "A0"."GROUP"
+                        AND "M"."LATEST_VERSION" <> "A0"."VERSION"
+                        AND "A0"."PURL" LIKE (('pkg:' || LOWER("M"."REPOSITORY_TYPE")) || '/%') ESCAPE E'\\\\'))
+                    """;
         }
         if (onlyDirect) {
-            querySring +=
-                    " && this.project.directDependencies.matches('%\"uuid\":\"'+this.uuid+'\"%') "; // only direct dependencies
+            queryString +=
+                    """
+                       AND "B0"."DIRECT_DEPENDENCIES" LIKE (('%' || "A0"."UUID") || '%') ESCAPE E'\\\\'
+                    """;
         }
-        final Query<Component> query = pm.newQuery(querySring);
-        query.getFetchPlan().setMaxFetchDepth(2);
         if (orderBy == null) {
-            query.setOrdering("name asc, version desc, id asc");
-        }
-        if (filter != null) {
-            final String filterString = ".*" + filter.toLowerCase() + ".*";
-            result = execute(query, project, filterString);
+            queryString +=
+                    """
+                        ORDER BY "NAME",
+                        "VERSION" DESC
+                    """;
         } else {
-            result = execute(query, project);
+            if (orderBy.equalsIgnoreCase("componentMetaInformation.publishedDate")) {
+                queryString +=
+                        """
+                            ORDER BY "PUBLISHED_AT"
+                        """;
+            }
+            if (orderBy.equalsIgnoreCase("version")) {
+                queryString +=
+                        """
+                            ORDER BY "VERSION"
+                        """;
+            }
+            if (orderBy.equalsIgnoreCase("name")) {
+                queryString +=
+                        """
+                            ORDER BY "NAME"
+                        """;
+            }
+            if (orderBy.equalsIgnoreCase("componentMetaInformation.integrityMatchStatus")) {
+                queryString +=
+                        """
+                            ORDER BY "INTEGRITY_CHECK_STATUS"
+                        """;
+            }
```

Comment:
In the UI it is also (currently) possible to sort by:

* Group
* Risk score

So at least those need to be added here as well.